### PR TITLE
parser: support heredoc format

### PR DESCRIPTION
**Pull request by m-ildefons**
- Add heredoc format support for everything that parses arguments (e.g.
  RUN).
- Separate out unit tests for RUN instructions into their own file
- Minor version bump

fixes: https://github.com/hadolint/hadolint/issues/682

Screenshot from 2021-08-07 13:11:40

### Notes:
This feature may require a bit more thought than what I've put into it so far. In particular I have these concerns:
- Do the unit tests cover enough cases or are there cases I haven't covered?
- Can this break parsing other instructions with arguments (CMD, HEALTHCHECK, etc)?
- What happens if a heredoc is not a shell script? Does this work robust enough in cases like:
```
RUN <<EOF
#/usr/bin/env python
print("foo")
EOF
```

### Todo:
- [x] Support markers other than `EOF`
- [x] Heredocs for `COPY` (https://www.docker.com/blog/introduction-to-heredocs-in-dockerfiles/) (including multiple heredocs)
- [x] Heredocs with redirects to file:

```
RUN <<EOF > /foo
echo foo
EOF
```
- [x] Heredocs intermixed with commands
```
RUN python <<EOF
print("foo")
EOF
```

### CODE REVIEW

File: src/Language/Docker/Parser/Copy.hs [hunk 1]
```diff
@@ -66,6 +67,15 @@ parseAdd = do
                   c : _ -> c
       fileList "ADD" (\src dest -> Add (AddArgs src dest cho chm))
 
+heredocList :: (?esc :: Char) =>
+               (NonEmpty SourcePath -> TargetPath -> Instruction Text) ->
+               Parser (Instruction Text)
+heredocList constr = do
+  sources <- spaceSep1 heredocMarker
+  target <- untilEol "target path"
+  void $ heredocContent (last sources)
```
Comment (lorenzo) [inline src/Language/Docker/Parser/Copy.hs:76]: Even when it is obviously correct here, the use of the partial fiction `last` makes me uncomfortable. Can you use pattern matching and fail if the list is empty?

File: src/Language/Docker/Parser/Prelude.hs [hunk 2]
```diff
@@ -126,15 +131,71 @@ natural = L.decimal <?> "positive number"
 commaSep :: (?esc :: Char) => Parser a -> Parser [a]
 commaSep p = sepBy (p <* whitespace) (symbol ",")
 
+spaceSep1 :: Parser a -> Parser [a]
+spaceSep1 p = sepEndBy1 p onlySpaces
+
+-- | Not this is just an alias for compatibility
```
Comment (lorenzo) [inline src/Language/Docker/Parser/Prelude.hs:137]: Not => Note

File: src/Language/Docker/Parser/Copy.hs (hunk 1 above)
Comment (m-ildefons) [reply src/Language/Docker/Parser/Copy.hs]: Pattern matching in this case really is not improving all that much, in particular the readability suffers quite a bit IMO. But then again I understand your concerns about `last`.

File: src/Language/Docker/Parser/Prelude.hs (hunk 2 above)
Comment (m-ildefons) [reply src/Language/Docker/Parser/Prelude.hs]: Good spot :-)

**Comment by m-ildefons** [comment]
@lorenzo I've fixed the usage of `last`. Could you please take another look at this? I'd really like to see heredoc support come to Hadolint.

**Comment by lorenzo** [comment]
sorry for the late review. Great job!

**Pull request merged**